src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/simple/SimpleJoinOrderingQueryOptimizer.java

```diff
@@ -41,14 +48,26 @@ public PhysicalPlan optimize( final LogicalPlan initialPlan ) throws QueryOptimi
     }
 
     public PhysicalPlan optimizePlan( final PhysicalPlan plan ) throws QueryOptimizationException {
+        final PhysicalPlan[] optSubPlans = getOptimizedSubPlans(plan);
+
         if ( hasMultiwayJoinAsRoot(plan) ){
-            final PhysicalPlan[] subplans = getOptimizedSubPlans(plan);
-            return joinPlanOptimizer.determineJoinPlan(subplans);
+            return joinPlanOptimizer.determineJoinPlan(optSubPlans);
         }
-        else {
-        	// TODO incorrect!! There may be multiway joins inside the subplans of the given plan.
+        else if (hasNullaryOpAsRoot(plan)){
             return plan;
         }
+        else if ( hasUnaryOpAsRoot(plan) ){
+            final PhysicalPlan newChild = optimizePlan(optSubPlans[0]);
```

Comment:
It is not necessary to call `optimizePlan` for `optSubPlans[0]`.  The new subplans in `optSubPlans` have already been optimized recursively. That's what the method `getOptimizedSubPlans` (as called above) is doing.